### ISSUE TITLE
OPA docker container fails to run as server in Azure IOT Edge module

### ISSUE THREAD

**Issue by sree1026** [open-policy-agent/opa#5663]
## Short description 
We tried running OPA as a server by building the docker image using openpolicyagent/opa:latest-rootless as base image and ran the container on the Azure IoT Edge device.

While running locally on our development PC which is windows machine or running on Azure IoT Edge simulator, we did not face any issue with OPA.

But while running on Azure IoT Edge device, the container immediately stopped running throwing error as below:

``` runtime/cgo: pthread_create failed: Operation not permitted
SIGABRT: abort
PC=0x7f0cd8480b8c m=0 sigcode=18446744073709551610

goroutine 0 [idle]:
runtime: g 0: unknown pc 0x7f0cd8480b8c
stack: frame={sp:0x7ffdb1e3cca0, fp:0x0} stack=[0x7ffdb163e120,0x7ffdb1e3d130)
0x00007ffdb1e3cba0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cbb0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cbc0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cbd0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cbe0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cbf0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc00:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc10:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc20:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc30:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc40:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc50:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc60:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc70:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc80:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc90:  0x0000000000000000  0x00007f0cd8480b7e
0x00007ffdb1e3cca0: <0x0000000000000000  0x00007f0cd84eaf54
0x00007ffdb1e3ccb0:  0x0000000000000006  0x00007f0cd83f8c40
0x00007ffdb1e3ccc0:  0x0000000003da02e0  0x0000000000000178
0x00007ffdb1e3ccd0:  0x0000000002dac040  0x00007f0cd8434402
0x00007ffdb1e3cce0:  0x00007f0cd85c2e70  0x00007f0cd841f464
0x00007ffdb1e3ccf0:  0x00007f0cd85be600  0x0000000000000020
0x00007ffdb1e3cd00:  0x0000000000000d68  0x00007f0cd84764e0
0x00007ffdb1e3cd10:  0x0000000000000000  0x0000000000000001
0x00007ffdb1e3cd20:  0x000000000000000a  0x00007f0cd83f8c40
0x00007ffdb1e3cd30:  0x0000000000000178  0x0000000002dac040
0x00007ffdb1e3cd40:  0x00007f0cb108735b  0x00007f0cd8478059
0x00007ffdb1e3cd50:  0x00007f0cd85c2680  0x00007f0cd8478463
0x00007ffdb1e3cd60:  0x00007f0cd85c2680  0x000000000000000a
0x00007ffdb1e3cd70:  0x00007f0cd83f8c40  0x00007f0cd8473c92
0x00007ffdb1e3cd80:  0x00007f0cd85c2840  0x000000000220a51a
0x00007ffdb1e3cd90:  0x00007f0cd85c2840  0x00007f0cd85c2840
runtime: g 0: unknown pc 0x7f0cd8480b8c
stack: frame={sp:0x7ffdb1e3cca0, fp:0x0} stack=[0x7ffdb163e120,0x7ffdb1e3d130)
0x00007ffdb1e3cba0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cbb0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cbc0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cbd0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cbe0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cbf0:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc00:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc10:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc20:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc30:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc40:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc50:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc60:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc70:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc80:  0x0000000000000000  0x0000000000000000
0x00007ffdb1e3cc90:  0x0000000000000000  0x00007f0cd8480b7e
0x00007ffdb1e3cca0: <0x0000000000000000  0x00007f0cd84eaf54
0x00007ffdb1e3ccb0:  0x0000000000000006  0x00007f0cd83f8c40
0x00007ffdb1e3ccc0:  0x0000000003da02e0  0x0000000000000178
0x00007ffdb1e3ccd0:  0x0000000002dac040  0x00007f0cd8434402
0x00007ffdb1e3cce0:  0x00007f0cd85c2e70  0x00007f0cd841f464
0x00007ffdb1e3ccf0:  0x00007f0cd85be600  0x0000000000000020
0x00007ffdb1e3cd00:  0x0000000000000d68  0x00007f0cd84764e0
0x00007ffdb1e3cd10:  0x0000000000000000  0x0000000000000001
0x00007ffdb1e3cd20:  0x000000000000000a  0x00007f0cd83f8c40
0x00007ffdb1e3cd30:  0x0000000000000178  0x0000000002dac040
0x00007ffdb1e3cd40:  0x00007f0cb108735b  0x00007f0cd8478059
0x00007ffdb1e3cd50:  0x00007f0cd85c2680  0x00007f0cd8478463
0x00007ffdb1e3cd60:  0x00007f0cd85c2680  0x000000000000000a
0x00007ffdb1e3cd70:  0x00007f0cd83f8c40  0x00007f0cd8473c92
0x00007ffdb1e3cd80:  0x00007f0cd85c2840  0x000000000220a51a
0x00007ffdb1e3cd90:  0x00007f0cd85c2840  0x00007f0cd85c2840

goroutine 1 [running]:
runtime.systemstack_switch()
        /usr/local/go/src/runtime/asm_amd64.s:459 fp=0xc000162780 sp=0xc000162778 pc=0x6a6000
runtime.main()
        /usr/local/go/src/runtime/proc.go:170 +0x6d fp=0xc0001627e0 sp=0xc000162780 pc=0x6762ed
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc0001627e8 sp=0xc0001627e0 pc=0x6a8221

rax    0x0
rbx    0x1
rcx    0x7f0cd8480b8c
rdx    0x6
rdi    0x1
rsi    0x1
rbp    0x7f0cd83f8c40
rsp    0x7ffdb1e3cca0
r8     0x0
r9     0x73
r10    0x8
r11    0x246
r12    0x6
r13    0x178
r14    0x2dac040
r15    0x7f0cb108735b
rip    0x7f0cd8480b8c
rflags 0x246
cs     0x33
fs     0x0
gs     0x0
runtime/cgo: pthread_create failed: Operation not permitted
SIGABRT: abort
PC=0x7f486ff85b8c m=0 sigcode=18446744073709551610

goroutine 0 [idle]:
runtime: g 0: unknown pc 0x7f486ff85b8c
stack: frame={sp:0x7ffcbabe01f0, fp:0x0} stack=[0x7ffcba3e1670,0x7ffcbabe0680)
0x00007ffcbabe00f0:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0100:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0110:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0120:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0130:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0140:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0150:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0160:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0170:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0180:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0190:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe01a0:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe01b0:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe01c0:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe01d0:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe01e0:  0x0000000000000000  0x00007f486ff85b7e
0x00007ffcbabe01f0: <0x0000000000000000  0x00007f486ffeff54
0x00007ffcbabe0200:  0x0000000000000006  0x00007f486fefdc40
0x00007ffcbabe0210:  0x00000000033212e0  0x0000000000000178
0x00007ffcbabe0220:  0x0000000002dac040  0x00007f486ff39402
0x00007ffcbabe0230:  0x00007f48700c7e70  0x00007f486ff24464
0x00007ffcbabe0240:  0x00007f48700c3600  0x0000000000000020
0x00007ffcbabe0250:  0x0000000000000d68  0x00007f486ff7b4e0
0x00007ffcbabe0260:  0x0000000000000000  0x0000000000000001
0x00007ffcbabe0270:  0x000000000000000a  0x00007f486fefdc40
0x00007ffcbabe0280:  0x0000000000000178  0x0000000002dac040
0x00007ffcbabe0290:  0x00007f4848b8c35b  0x00007f486ff7d059
0x00007ffcbabe02a0:  0x00007f48700c7680  0x00007f486ff7d463
0x00007ffcbabe02b0:  0x00007f48700c7680  0x000000000000000a
0x00007ffcbabe02c0:  0x00007f486fefdc40  0x00007f486ff78c92
0x00007ffcbabe02d0:  0x00007f48700c7840  0x000000000220a51a
0x00007ffcbabe02e0:  0x00007f48700c7840  0x00007f48700c7840
runtime: g 0: unknown pc 0x7f486ff85b8c
stack: frame={sp:0x7ffcbabe01f0, fp:0x0} stack=[0x7ffcba3e1670,0x7ffcbabe0680)
0x00007ffcbabe00f0:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0100:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0110:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0120:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0130:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0140:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0150:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0160:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0170:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0180:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe0190:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe01a0:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe01b0:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe01c0:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe01d0:  0x0000000000000000  0x0000000000000000
0x00007ffcbabe01e0:  0x0000000000000000  0x00007f486ff85b7e
0x00007ffcbabe01f0: <0x0000000000000000  0x00007f486ffeff54
0x00007ffcbabe0200:  0x0000000000000006  0x00007f486fefdc40
0x00007ffcbabe0210:  0x00000000033212e0  0x0000000000000178
0x00007ffcbabe0220:  0x0000000002dac040  0x00007f486ff39402
0x00007ffcbabe0230:  0x00007f48700c7e70  0x00007f486ff24464
0x00007ffcbabe0240:  0x00007f48700c3600  0x0000000000000020
0x00007ffcbabe0250:  0x0000000000000d68  0x00007f486ff7b4e0
0x00007ffcbabe0260:  0x0000000000000000  0x0000000000000001
0x00007ffcbabe0270:  0x000000000000000a  0x00007f486fefdc40
0x00007ffcbabe0280:  0x0000000000000178  0x0000000002dac040
0x00007ffcbabe0290:  0x00007f4848b8c35b  0x00007f486ff7d059
0x00007ffcbabe02a0:  0x00007f48700c7680  0x00007f486ff7d463
0x00007ffcbabe02b0:  0x00007f48700c7680  0x000000000000000a
0x00007ffcbabe02c0:  0x00007f486fefdc40  0x00007f486ff78c92
0x00007ffcbabe02d0:  0x00007f48700c7840  0x000000000220a51a
0x00007ffcbabe02e0:  0x00007f48700c7840  0x00007f48700c7840

goroutine 1 [running]:
runtime.systemstack_switch()
        /usr/local/go/src/runtime/asm_amd64.s:459 fp=0xc000162780 sp=0xc000162778 pc=0x6a6000
runtime.main()
        /usr/local/go/src/runtime/proc.go:170 +0x6d fp=0xc0001627e0 sp=0xc000162780 pc=0x6762ed
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:1594 +0x1 fp=0xc0001627e8 sp=0xc0001627e0 pc=0x6a8221

rax    0x0
rbx    0x1
rcx    0x7f486ff85b8c
rdx    0x6
rdi    0x1
rsi    0x1
rbp    0x7f486fefdc40
rsp    0x7ffcbabe01f0
r8     0x0
r9     0x73
r10    0x8
r11    0x246
r12    0x6
r13    0x178
r14    0x2dac040
r15    0x7f4848b8c35b
rip    0x7f486ff85b8c
rflags 0x246
cs     0x33
fs     0x0
gs     0x0
runtime/cgo: pthread_create failed: Operation not permitted
SIGABRT: abort
PC=0x7fcb23adbb8c m=0 sigcode=18446744073709551610

goroutine 0 [idle]:
runtime: g 0: unknown pc 0x7fcb23adbb8c
stack: frame={sp:0x7fff3dae5450, fp:0x0} stack=[0x7fff3d2e68d0,0x7fff3dae58e0)
0x00007fff3dae5350:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5360:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5370:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5380:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5390:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53a0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53b0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53c0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53d0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53e0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53f0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5400:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5410:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5420:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5430:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5440:  0x0000000000000000  0x00007fcb23adbb7e
0x00007fff3dae5450: <0x0000000000000000  0x00007fcb23b45f54
0x00007fff3dae5460:  0x0000000000000006  0x00007fcb23a53c40
0x00007fff3dae5470:  0x0000000003b802e0  0x0000000000000178
0x00007fff3dae5480:  0x0000000002dac040  0x00007fcb23a8f402
0x00007fff3dae5490:  0x00007fcb23c1de70  0x00007fcb23a7a464
0x00007fff3dae54a0:  0x00007fcb23c19600  0x0000000000000020
0x00007fff3dae54b0:  0x0000000000000d68  0x00007fcb23ad14e0
0x00007fff3dae54c0:  0x0000000000000000  0x0000000000000001
0x00007fff3dae54d0:  0x000000000000000a  0x00007fcb23a53c40
0x00007fff3dae54e0:  0x0000000000000178  0x0000000002dac040
0x00007fff3dae54f0:  0x00007fcafc6e235b  0x00007fcb23ad3059
0x00007fff3dae5500:  0x00007fcb23c1d680  0x00007fcb23ad3463
0x00007fff3dae5510:  0x00007fcb23c1d680  0x000000000000000a
0x00007fff3dae5520:  0x00007fcb23a53c40  0x00007fcb23acec92
0x00007fff3dae5530:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5540:  0x00007fcb23c1d840  0x00007fcb23c1d840
runtime: g 0: unknown pc 0x7fcb23adbb8c
stack: frame={sp:0x7fff3dae5450, fp:0x0} stack=[0x7fff3d2e68d0,0x7fff3dae58e0)
0x00007fff3dae5350:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5360:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5370:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5380:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5390:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53a0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53b0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53c0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53d0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53e0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae53f0:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5400:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5410:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5420:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5430:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5440:  0x0000000000000000  0x00007fcb23adbb7e
0x00007fff3dae5450: <0x0000000000000000  0x00007fcb23b45f54
0x00007fff3dae5460:  0x0000000000000006  0x00007fcb23a53c40
0x00007fff3dae5470:  0x0000000003b802e0  0x0000000000000178
0x00007fff3dae5480:  0x0000000002dac040  0x00007fcb23a8f402
0x00007fff3dae5490:  0x00007fcb23c1de70  0x00007fcb23a7a464
0x00007fff3dae54a0:  0x00007fcb23c19600  0x0000000000000020
0x00007fff3dae54b0:  0x0000000000000d68  0x00007fcb23ad14e0
0x00007fff3dae54c0:  0x0000000000000000  0x0000000000000001
0x00007fff3dae54d0:  0x000000000000000a  0x00007fcb23a53c40
0x00007fff3dae54e0:  0x0000000000000178  0x0000000002dac040
0x00007fff3dae54f0:  0x00007fcafc6e235b  0x00007fcb23ad3059
0x00007fff3dae5500:  0x00007fcb23c1d680  0x00007fcb23ad3463
0x00007fff3dae5510:  0x00007fcb23c1d680  0x000000000000000a
0x00007fff3dae5520:  0x00007fcb23a53c40  0x00007fcb23acec92
0x00007fff3dae5530:  0x0000000000000000  0x0000000000000000
0x00007fff3dae5540:  0x00007fcb23c1d840  0x00007fcb23c1d840 
```

Examples:
* OPA version: [(openpolicyagent/opa:latest-rootless)](https://hub.docker.com/layers/openpolicyagent/opa/latest-rootless/images/sha256-b8d2ca87f0241531433d106473bbe3661b7c9be735c447daefa164f2c3942b8d?context=explore)
* Docker File
```
FROM openpolicyagent/opa:latest-rootless

WORKDIR /src

COPY /policy/policy.rego /src/policy/

COPY /data/data.json /src/data/

EXPOSE 8181

CMD ["run", "--server", "--log-format=json-pretty", "policy/policy.rego", "data/data.json"] 
```
* Azure IoT Edge device:
NAME="Ubuntu"
VERSION="18.04.5 LTS"

## Steps To Reproduce
1. Build the docker image
2. Deploy the docker image on Azure IoT Edge device

## Expected behavior
OPA container should run without any issues

## Additional context
Similar issue was reported on a different software [here](https://github.com/containers/skopeo/issues/1501)

The workaround mentioned was to modify "seccomp" while running the docker container and we tried doing the same i.e. we made
"seccomp=unconfined" which resolved the issue. 

I'm not sure how did it resolve the issue, hence posting it here to know if this is indeed the recommended way to fix the issue or if there is any other better way and also to know more about why the issue happened in first place

Please let me know if any further details is required.

**Comment by ashutosh-narkar** [comment]
Could this be related to the docker version as suggested in the link you shared. Also I'd look how the Azure IoT Edge simulator env differs from the device. Have you tried using `openpolicyagent/opa:latest-static` image?

**Comment by ashutosh-narkar** [comment]
Closing this. Feel free to re-open if issue still exists.

**Comment by sree1026** [comment]
Thanks a lot @ashutosh-narkar and apologies for very much delayed response. I tried using "latest-static" image of OPA and it seems to work fine. The container was running for few days and we did not face any problems as reported earlier by me.

I would like to know how static image is working fine whereas the actual image is throwing some runtime errors. Please share if you have any leads on why the error is being thrown.

Also, is there rootless static image available for opa? Current static image is running as root and its throwing a warning to use rootless and I tried to pull `openpolicyagent/opa:latest-static-rootless` , seems like its not available.

**Comment by ashutosh-narkar** [comment]
> Also, is there rootless static image available for opa? Current static image is running as root and its throwing a warning to use rootless and I tried to pull openpolicyagent/opa:latest-static-rootless , seems like its not available.

All published OPA images now (with the `0.50.0` release) run with a non-root uid/gid.  The uid:gid is set to 1000:1000 for all images. This is true for `latest-static` image variant.

**Comment by sree1026** [comment]
Great! Thanks for the quick response.
